servers/graphql-kotlin-server/src/main/kotlin/com/expediagroup/graphql/server/execution/GraphQLRequestHandler.kt

```diff
@@ -16,37 +16,139 @@
      * This should only be used for queries and mutations.
      * Subscriptions require more specific server logic and will need to be handled separately.
      */
-    open suspend fun executeRequest(request: GraphQLRequest, context: GraphQLContext? = null, graphQLContext: Map<*, Any> = emptyMap<Any, Any>()): GraphQLResponse<*> {
-        // We should generate a new registry for every request
-        val dataLoaderRegistry = dataLoaderRegistryFactory?.generate()
-        val executionInput = request.toExecutionInput(context, dataLoaderRegistry, graphQLContext)
+    open suspend fun executeRequest(
+        graphQLRequest: GraphQLServerRequest,
+        context: GraphQLContext? = null,
+        graphQLContext: Map<*, Any> = emptyMap<Any, Any>()
+    ): GraphQLServerResponse =
+        when (graphQLRequest) {
+            is GraphQLRequest -> {
+                execute(graphQLRequest, context, graphQLContext, dataLoaderRegistryFactory?.generate())
+            }
+            is GraphQLBatchRequest -> {
+                if (graphQLRequest.containsMutation()) {
+                    executeBatchSequentially(graphQLRequest, context, graphQLContext)
+                } else {
+                    executeBatchConcurrently(graphQLRequest, context, graphQLContext)
+                }
+            }
+        }
 
+    private suspend fun execute(
+        graphQLRequest: GraphQLRequest,
+        context: GraphQLContext? = null,
+        graphQLContext: Map<*, Any> = emptyMap<Any, Any>(),
+        dataLoaderRegistry: KotlinDataLoaderRegistry? = null
+    ): GraphQLResponse<*> {
+        val executionInput = graphQLRequest.toExecutionInput(context, dataLoaderRegistry, graphQLContext)
         return try {
             graphQL.executeAsync(executionInput).await().toGraphQLResponse()
         } catch (exception: Exception) {
             val error = exception.toGraphQLError()
             GraphQLResponse<Any?>(errors = listOf(error.toGraphQLKotlinType()))
         }
     }
+
+    private suspend fun executeBatchSequentially(
+        request: GraphQLBatchRequest,
+        context: GraphQLContext?,
+        graphQLContext: Map<*, Any>
+    ): GraphQLBatchResponse {
+        val dataLoaderRegistry = dataLoaderRegistryFactory?.generate()
+        return GraphQLBatchResponse(
+            request.requests.map {
+                execute(it, context, graphQLContext, dataLoaderRegistry)
+            }
+        )
+    }
+
+    private suspend fun executeBatchConcurrently(
+        request: GraphQLBatchRequest,
+        context: GraphQLContext?,
+        graphQLContext: Map<*, Any>
+    ): GraphQLBatchResponse {
+        val batchContext = getBatchContext(request, dataLoaderRegistryFactory?.generate())
+        val batchGraphQLContext = graphQLContext + batchContext
+        val responses = supervisorScope {
+            request.requests.map {
+                async {
+                    execute(it, context, batchGraphQLContext)
```

Comment:
depending whether there is batch instrumentation or not you may have to provide the dataloader instance either in graphql context (batch instrumentation enabled) OR as part of the execution input (no batch)